src/Compilers/CSharp/Test/Emit/PDB/PDBSourceLinkTests.cs

```diff
@@ -0,0 +1,154 @@
+    {
+        Console.WriteLine();
+    }
+}
+";
+            var sourceLinkBlob = Encoding.UTF8.GetBytes(@"
+{
+  ""documents"": {
+     ""f:/build/*"" : ""https://raw.githubusercontent.com/my-org/my-project/1111111111111111111111111111111111111111/*""
+  }
+}
+");
+
+            var c = CreateStandardCompilation(Parse(source, "f:/build/foo.cs"), options: TestOptions.DebugDll);
+
+            var pdbStream = new MemoryStream();
+            c.EmitToArray(EmitOptions.Default.WithDebugInformationFormat(format), pdbStream: pdbStream, sourceLinkStream: new MemoryStream(sourceLinkBlob));
+            pdbStream.Position = 0;
+
+            var actualData = PdbValidation.GetSourceLinkData(pdbStream);
+            AssertEx.Equal(sourceLinkBlob, actualData);
+        }
+
+        [Fact]
+        public void SourceLink_Embedded()
+        {
+            string source = @"
+using System;
+
+class C
+{
+    public static void Main()
+    {
+        Console.WriteLine();
+    }
+}
+";
+            var sourceLinkBlob = Encoding.UTF8.GetBytes(@"
+{
+  ""documents"": {
+     ""f:/build/*"" : ""https://raw.githubusercontent.com/my-org/my-project/1111111111111111111111111111111111111111/*""
+  }
+}
+");
+            var c = CreateStandardCompilation(Parse(source, "f:/build/foo.cs"), options: TestOptions.DebugDll);
+
+            var peBlob = c.EmitToArray(EmitOptions.Default.WithDebugInformationFormat(DebugInformationFormat.Embedded), sourceLinkStream: new MemoryStream(sourceLinkBlob));
+
+            using (var peReader = new PEReader(peBlob))
+            {
+                var embeddedEntry = peReader.ReadDebugDirectory().Single(e => e.Type == DebugDirectoryEntryType.EmbeddedPortablePdb);
+
+                using (var embeddedMetadataProvider = peReader.ReadEmbeddedPortablePdbDebugDirectoryData(embeddedEntry))
+                {
+                    var pdbReader = embeddedMetadataProvider.GetMetadataReader();
+
+                    var actualBlob =
+                        (from cdiHandle in pdbReader.GetCustomDebugInformation(EntityHandle.ModuleDefinition)
+                         let cdi = pdbReader.GetCustomDebugInformation(cdiHandle)
+                         where pdbReader.GetGuid(cdi.Kind) == PortableCustomDebugInfoKinds.SourceLink
+                         select pdbReader.GetBlobBytes(cdi.Value)).Single();
+
+                    AssertEx.Equal(sourceLinkBlob, actualBlob);
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData(DebugInformationFormat.Pdb)]
+        [InlineData(DebugInformationFormat.Embedded)]
+        [InlineData(DebugInformationFormat.PortablePdb)]
+        public void SourceLink_Errors(DebugInformationFormat format)
+        {
+            string source = @"
+using System;
+
+class C
+{
+    public static void Main()
+    {
+        Console.WriteLine();
+    }
+}
+";
+            var sourceLinkStream = new TestStream(canRead: true, readFunc: (_, __, ___) => { throw new Exception("Error!"); });
```

Comment:
+1

---
In reply to: [113744881](https://github.com/dotnet/roslyn/pull/18539#discussion_r113744881) [](ancestors = 113744881)